core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
+
+        @Override
+        public CharSequence getStr(Record rec) {
+            return getStr0(rec, sink);
+        }
+
+        @Override
+        public CharSequence getStrB(Record rec) {
+            return getStr0(rec, sinkB);
+        }
+
+        @Override
+        public void getStr(Record rec, CharSink sink) {
+            getStr0(rec, sink);
+        }
+
+        @Nullable
+        private <S extends CharSink> S getStr0(Record rec, S sink) {
+            var str = strFunc.getStr(rec);
+            var delimiter = delimiterFunc.getStr(rec);
+            var index = getIndex(rec);
+            if (str == null || delimiter == null || index == Numbers.INT_NaN) {
+                return null;
+            }
+            if (index == 0) {
+                return sink;
+            }
+
+            int start;
+            int end;
+            if (index > 0) {
+                if (index == 1) {
+                    start = 0;
+                } else {
+                    start = Chars.indexOf(str, 0, str.length(), delimiter, index - 1);
+                    if (start == -1) {
+                        return sink;
+                    }
+                    start += delimiter.length();
+                }
+
+                end = Chars.indexOf(str, start, str.length(), delimiter, index);
```

Comment:
nit: `split_part()` function could be implemented in a single pass over the input string. Currently, there are two passes.

Reply:
That's a valid point. Marking this one as resolved then.